src/app/core/services/kafka/schema.service.ts

```diff
@@ -153,17 +154,34 @@ export class SchemaService {
       })
   }
 
-  getKafkaTopic(specifications: any[] | null, name, avsc): Promise<any> {
-    try {
-      const type = name.toLowerCase()
-      const defaultTopic = `${avsc}_${name}`
-      if (specifications) {
-        const spec = specifications.find(t => t.type.toLowerCase() == type)
-        return Promise.resolve(spec && spec.topic ? spec.topic : defaultTopic)
+  getKafkaTopic(specifications: any[] | null, name, avsc, topics: string[] | null): Promise<any> {
+    const type = name.toLowerCase()
+
+    if (specifications) {
+      const spec = specifications.find(t => t.type.toLowerCase() == type)
+      if (spec && spec.topic && this.topicExists(spec.topic, topics)) {
+        return Promise.resolve(spec.topic)
       }
+    }
+    const questionnaireTopic = `${avsc}_${name}`
+    if (this.topicExists(questionnaireTopic, topics)) {
+      return Promise.resolve(questionnaireTopic)
+    }
+    const defaultTopic = this.GENERAL_TOPIC;
+    if (this.topicExists(defaultTopic, topics)) {
       return Promise.resolve(defaultTopic)
-    } catch (e) {
-      return Promise.reject('Failed to get kafka topic')
+    }
+
+    return Promise.reject(`No suitable topic found on server for questionnaire ${name}`)
+  }
+
+  private topicExists(topic: string, topics: string[] | null) {
+    if (!topics || topics.includes(topic)) {
```

Comment:
should this be `&&` ?

Reply:
No this is intentional, it means: if topics _could not be_ fetched **or** topics _were_ fetched and they include the current topic. Then the topic is valid.

Reply:
Then how do we prevent a scenario of the app attempting to send a response without fetching and validating the current topic at all? At least we want to be sure that the `questionnaire_response` is available on the server. Am I right?

Reply:
If the topic cannot be validated to exist beforehand, the app will just attempt to send to a non-existing topic. It will fail, and then at a later time it will try again and hopefully the list of topics can then be successfully fetched. If it succeeded, then an additional check for the list of topics would not change anything.